pkg/operator/ceph/pool/controller.go

```diff
@@ -18,155 +18,186 @@ limitations under the License.
+		if err != nil {
+			logger.Errorf("failed to remove pool finalizer. %v", err)
+			return cephclient.ImmediateRetryResult, err
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ProcessingStatus, c.context)
 
-	// if the pool is modified, allow the pool to be created if it wasn't already
-	logger.Infof("updating pool %q", pool.Name)
-	if err := createPool(c.context, pool); err != nil {
-		logger.Errorf("failed to create (modify) pool %q. %v", pool.ObjectMeta.Name, err)
-		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.FailedStatus, c.context)
-		return
+	// CREATE or UPDATE
+	err = r.reconcile(cephBlockPool)
+	if err != nil {
+		logger.Errorf("failed to reconcile. %v", err)
+		updateCephBlockPoolStatus(cephBlockPool.GetName(), cephBlockPool.GetNamespace(), k8sutil.ReconcileFailedStatus, r.context)
+		return cephclient.ImmediateRetryResult, err
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ReadyStatus, c.context)
-}
 
-// ParentClusterChanged determines wether or not a CR update has been sent
-func (c *PoolController) ParentClusterChanged(cluster cephv1.ClusterSpec, clusterInfo *cephconfig.ClusterInfo, isUpgrade bool) {
-	logger.Debugf("No need to update the pool after the parent cluster changed")
+	return reconcile.Result{}, nil
 }
 
-func poolChanged(old, new cephv1.PoolSpec) bool {
-	if old.Replicated.Size != new.Replicated.Size {
-		logger.Infof("pool replication changed from %d to %d", old.Replicated.Size, new.Replicated.Size)
-		return true
-	}
-	return false
-}
+func (r *ReconcileCephBlockPool) reconcile(pool *cephv1.CephBlockPool) error {
+	// Create pools
+	if pool.Status.Phase != k8sutil.ReadyStatus {
+		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ProcessingStatus, r.context)
+		err := createPool(r.context, pool)
+		if err != nil {
+			updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.FailedStatus, r.context)
+			return errors.Wrapf(err, "failed to create pool %q.", pool.GetName())
+		}
 
-func (c *PoolController) onDelete(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Deleting pools for an external ceph cluster is disabled because no Ceph image is specified")
-		return
-	}
+		// Set Ready status
+		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ReadyStatus, r.context)
 
-	pool, err := getPoolObject(obj)
-	if err != nil {
-		logger.Errorf("failed to get pool object. %v", err)
-		return
+		// Let's return here so that on the initial creation we don't check for update right away
+		return nil
 	}
-	if err := deletePool(c.context, pool); err != nil {
-		logger.Errorf("failed to delete pool %q. %v", pool.ObjectMeta.Name, err)
+
+	// Verify whether the pool needs to be updated
+	if pool.Status.Phase == k8sutil.ReadyStatus {
```

Comment:
Seems like this could either just be an `else` from the previous block, or no need to check the condition since we already returned after the status==ready block above.